modules/imgproc/test/test_stackblur.cpp

```diff
@@ -0,0 +1,336 @@
+                sumIn[ci] -= tmp;
+            }
+
+            dstPtr += CN;
+
+            if (xp < wm)
+            {
+                xp++;
+                srcPtr += CN;
+            }
+
+            ++sp;
+            if (sp >= stackLenW) sp = 0;
+        }
+    }
+
+    // Vertical direction
+    int hm = rowsImg - 1;
+    int widthElem = colsImg * CN;
+    int radiusH = ksize.height / 2;
+    int stackLenH = ksize.height;
+    const float mulH = 1.0f / (((float )radiusH + 1.0f) * ((float )radiusH + 1.0f));
+
+    for (int col = 0; col < widthElem; col++)
+    {
+        const T* srcPtr =dst.ptr<T>() + col;
+        float sum0 = 0;
+        float sumIn0 = 0;
+        float sumOut0 = 0;
+
+        T* stack0 = (T* )bufptr;
+        memset(bufptr, 0, bufSize);
+
+        for (int i = 0; i <= radiusH; i++)
+        {
+            T tmp = (T)(*srcPtr);
+            stack0[i] = tmp;
+            sum0 += tmp * (i + 1);
+            sumOut0 += tmp;
+        }
+
+        for (int i = 1; i <= radiusH; i++)
+        {
+            if (i <= hm) srcPtr += widthElem;
+            T tmp = (T)(*srcPtr);
+            stack0[i + radiusH] = tmp;
+            sum0 += tmp * (radiusH - i + 1);
+            sumIn0 += tmp;
+        }
+
+        int sp = radiusH;
+        int yp = radiusH;
+
+        if (yp > hm) yp = hm;
+
+        T* dstPtr = dst.ptr<T>() + col;
+        srcPtr = dst.ptr<T>(yp) + col;
+
+        const T* srcNew;
+
+        int stackStart = 0;
+
+        for (int i = 0; i < rowsImg; i++)
+        {
+            stackStart = sp + stackLenH - radiusH;
+            if (stackStart >= stackLenH) stackStart -= stackLenH;
+
+            *(dstPtr) = saturate_cast<T>(sum0 * mulH);
+            sum0 -= sumOut0;
+            sumOut0 -= stack0[stackStart];
+            srcNew = srcPtr;
+
+            if (yp < hm)
+                srcNew += widthElem;
+
+            stack0[stackStart] = *(srcNew);
+            sumIn0 += *(srcNew);
+            sum0 += sumIn0;
+
+            int sp1 = sp + 1;
+            sp1 &= -(sp1 < stackLenH);
+
+            sumOut0 += stack0[sp1];
+            sumIn0 -= stack0[sp1];
+
+            dstPtr += widthElem;
+
+            if (yp < hm)
+            {
+                yp++;
+                srcPtr += widthElem;
+            }
+
+            ++sp;
+            if (sp >= stackLenH) sp = 0;
+        }
+    }
+}
+
+void stackBlurRef(const Mat& img, Mat& dst, Size ksize)
+{
+    if(img.depth() == CV_8U)
+        _stackblurRef<uchar>(img, dst, ksize);
+    else if (img.depth() == CV_16S)
+        _stackblurRef<short>(img, dst, ksize);
+    else if (img.depth() == CV_16U)
+        _stackblurRef<ushort>(img, dst, ksize);
+    else if (img.depth() == CV_32F)
+        _stackblurRef<float>(img, dst, ksize);
+    else
+        CV_Error_( CV_StsNotImplemented,
+                   ("Unsupported Mat type in stackBlurRef, "
+                    "the supported formats are: CV_8U, CV_16U, CV_16S and CV_32F."));
+}
+
+TEST(Stackblur, regression)
```

Comment:
It makes sense to make the test parameterized and check only one combination at a time. The same regression and stackedBlur against GaussianBlur. It should be dedicated set of parameterized tests.